src/main/java/com/transloadit/sdk/Assembly.java

```
@@ -1,64 +1,100 @@
      */
     public Assembly(Transloadit transloadit, Steps steps, Map<String, File> files, Map<String, Object> options) {
         this.transloadit = transloadit;
         this.steps = steps;
         this.files = files;
         this.options = options;
-        tusClient = null;
+        tusClient = new TusClient();
+        tusURLStore = new TusURLMemoryStore();
+        uploads = new ArrayList<TusUpload>();
+        fileStreams = new HashMap<String, InputStream>();
     }
 
     /**
      * Adds a file to your assembly.
      *
      * @param file {@link File} the file to be uploaded.
-     * @param name {@link String} the name you the file to be given in transloadit
+     * @param name {@link String} the name you the file to be given in transloadit.
      */
-    public void addFile(File file, String name){
+    public void addFile(File file, String name) {
         files.put(name, file);
+
+        // remove duplicate key
+        if (fileStreams.containsKey(name)) {
+            fileStreams.remove(name);
+        }
     }
 
     /**
      * Adds a file to your assembly but automatically genarates the name of the file.
      *
      * @param file {@link File} the file to be uploaded.
      */
-    public void addFile(File file){
+    public void addFile(File file) {
         String name = "file";
+        files.put(normalizeDuplicateName(name), file);
+    }
 
-        for (int i = files.size(); files.containsKey(name); i++) {
-            name += "_" + i;
+    /**
+     * Adds a file to your assembly.
+     *
+     * @param inputStream {@link InputStream} the file to be uploaded.
+     * @param name {@link String} the name you the file to be given in transloadit.
+     */
+    public void addFile(InputStream inputStream, String name) {
+        fileStreams.put(name, inputStream);
+
+        // remove duplicate key
+        if (files.containsKey(name)) {
+            files.remove(name);
```

What about being explicit and throwing an exception here instead of silently removing the first file?

---

same as my last comment

---

Ok, makes sense. Would you mind adding this to the method's JavaDoc?